libraries/apollo-compiler/src/main/kotlin/com/apollographql/apollo3/compiler/internal/addRequiredFields.kt

```diff
@@ -75,16 +77,22 @@ private fun List<GQLSelection>.addRequiredFields(
 
   val selectionSet = this
 
-  val requiresTypename = when(addTypename) {
+  val requiresTypename = when (addTypename) {
     "ifPolymorphic" -> isRoot && isPolymorphic(schema, fragments, parentType)
     "ifFragments" -> {
       selectionSet.any { it is GQLFragmentSpread || it is GQLInlineFragment }
     }
+
     "ifAbstract" -> isRoot && schema.typeDefinition(parentType).isAbstract()
     "always" -> isRoot
     else -> error("Unknown addTypename option: $addTypename")
   }
-  val requiredFieldNames = schema.keyFields(parentType).toMutableSet()
+  val requiredFieldNames = if (schema.typeDefinition(parentType) is GQLUnionTypeDefinition) {
+    // Can't select any fields on unions other than __typename
+    mutableSetOf()
+  } else {
+    schema.keyFields(parentType).toMutableSet()
+  }
```

Comment:
Should we allow `@typePolicy` on unions in the first place?

```
// This feels out of place. SomeUnion has no `id` field.
extend union SomeUnion @typePolicy(keyFields: "id") 
```

Reply:
Yeah I was wondering about this :) We could forbid it too, it might be less confusing.

Reply:
For the record, we talked about this and decided to keep the directive definition untouched but add validation that the `keyFields` actually exists, which should make an error here.

Reply:
Turns out this validation already existed but wasn't performed on interfaces nor unions.